station-items-v0/src/main/java/net/modificationstation/stationapi/api/bonemeal/BonemealAPI.java

```diff
@@ -0,0 +1,104 @@
+package net.modificationstation.stationapi.api.bonemeal;
+
+import it.unimi.dsi.fastutil.objects.Reference2ObjectOpenHashMap;
+import net.minecraft.block.BlockBase;
+import net.minecraft.level.Level;
+import net.minecraft.level.structure.Structure;
+import net.modificationstation.stationapi.api.block.BlockState;
+import net.modificationstation.stationapi.api.registry.BlockRegistry;
+import net.modificationstation.stationapi.api.tag.TagKey;
+import net.modificationstation.stationapi.api.util.collection.WeightedList;
+import net.modificationstation.stationapi.api.util.math.Direction;
+
+import java.util.Map;
+import java.util.Random;
+
+public class BonemealAPI {
+	private static final Map<BlockState, WeightedList<Structure>> PLACERS = new Reference2ObjectOpenHashMap<>();
+	
+	public static void addPlant(BlockState ground, BlockState plant, int weight) {
+		addPlant(ground, new SimpleStateStructure(plant), weight);
+	}
+	
+	public static void addPlant(BlockState ground, Structure plant, int weight) {
+		PLACERS.computeIfAbsent(ground, g -> new WeightedList<>()).add(plant, weight);
+	}
+	
+	public static void addPlant(TagKey<BlockBase> ground, BlockState plant, int weight) {
+		BlockRegistry.INSTANCE.forEach(blockBase ->
+			blockBase.getStateManager().getStates().stream().filter(state -> state.isIn(ground)).forEach(state -> addPlant(state, plant, weight))
+		);
+	}
+	
+	public static void addPlant(TagKey<BlockBase> ground, Structure plant, int weight) {
+		BlockRegistry.INSTANCE.forEach(blockBase ->
+			blockBase.getStateManager().getStates().stream().filter(state -> state.isIn(ground)).forEach(state -> addPlant(state, plant, weight))
```

Comment:
Caching states from tags like this is probably not a very good idea, tags are reloadable after all.

Reply:
Well, original idea was to add plants to all blocks with tag, like "nether_ground" or similar things. I can adapt code to work with additional internal tag storage, that can solve this issue

Reply:
Done, now it should support both reloadable tags and states